src/prof_recent.c

```diff
@@ -506,33 +580,65 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
 	emitter_begin(&emitter);
 
 	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
-	prof_recent_alloc_assert_count(tsd);
+	ssize_t dump_count = prof_recent_alloc_dump_prepare_locked(tsd);
+	ssize_t dump_max = prof_recent_alloc_max_get(tsd);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 
-	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
-	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize,
+		&dump_max);
+	emitter_json_array_kv_begin(&emitter, "recent_alloc");
 
-	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
+	prof_recent_t *dump_last = NULL;
 
-	emitter_json_array_kv_begin(&emitter, "recent_alloc");
-	for (prof_recent_t *n = prof_recent_alloc_begin(tsd);
-	    n != prof_recent_alloc_end(tsd);
-	    n = prof_recent_alloc_next(tsd, n)) {
-		prof_recent_alloc_dump_node(&emitter, n);
+	buf_writer_set_allow_auto_flush(&buf_writer, false);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *dump_dummy = prof_recent_alloc_dummy_reserve;
+	for (prof_recent_t *node = dump_dummy->next; node != dump_dummy;
+	    node = node->next) {
+		buf_writer_commit(&buf_writer);
+		prof_recent_alloc_dump_node(&emitter, node);
+		if (buf_writer_overflow(&buf_writer)) {
+			buf_writer_revert(&buf_writer);
+			if (buf_writer_empty(&buf_writer)) {
+				buf_writer_set_allow_auto_flush(
+				    &buf_writer, true);
+				prof_recent_alloc_dump_node(
+				    &emitter, node);
+				buf_writer_set_allow_auto_flush(
+				    &buf_writer, false);
+			} else {
+				malloc_mutex_unlock(tsd_tsdn(tsd),
+				    &prof_recent_alloc_mtx);
+				buf_writer_flush(&buf_writer);
+				malloc_mutex_lock(tsd_tsdn(tsd),
+				    &prof_recent_alloc_mtx);
+			}
+		}
+		dump_last = node;
 	}
-	emitter_json_array_end(&emitter);
-
-	max = prof_recent_alloc_max_update(tsd, max);
-	assert(max == 0);
 	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	buf_writer_set_allow_auto_flush(&buf_writer, true);
 
+	emitter_json_array_end(&emitter);
 	emitter_end(&emitter);
 	buf_writer_terminate(tsd_tsdn(tsd), &buf_writer);
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	ssize_t new_max = prof_recent_alloc_dump_finish_locked(tsd, dump_last,
```

Comment:
Also, this one be move to right after the for loop?

Reply:
Good point.